File: app/src/main/java/md/fusionworks/lifehack/data/api/model/exchange_rates/mapper/BranchDataMapper.java
package md.fusionworks.lifehack.data.api.model.exchange_rates.mapper;

import java.util.ArrayList;
import java.util.List;
import md.fusionworks.lifehack.data.api.model.exchange_rates.Branch;
import md.fusionworks.lifehack.ui.exchange_rates.model.AddressModel;
import md.fusionworks.lifehack.ui.exchange_rates.model.BranchModel;
import md.fusionworks.lifehack.ui.exchange_rates.model.DayModel;
import md.fusionworks.lifehack.ui.exchange_rates.model.DistrictModel;
import md.fusionworks.lifehack.ui.exchange_rates.model.LocalityModel;
import md.fusionworks.lifehack.ui.exchange_rates.model.LocationModel;
import md.fusionworks.lifehack.ui.exchange_rates.model.ScheduleModel;

/**
 * Created by ungvas on 2/10/16.
 */
public class BranchDataMapper {

  public BranchModel transform(Branch branch) {

    LocalityModel localityModel =
        new LocalityModel(branch.getAddress().getDistrict().getLocality().getId(),
            branch.getAddress().getDistrict().getLocality().getName(),
            branch.getAddress().getDistrict().getLocality().getCode());
    DistrictModel districtModel = new DistrictModel(branch.getAddress().getDistrict().getId(),
        branch.getAddress().getDistrict().getName(), localityModel);
    AddressModel addressModel = new AddressModel(
        new LocationModel(branch.getAddress().getLocation().getLat(),
            branch.getAddress().getLocation().getLng()), branch.getAddress().getId(), districtModel,
        branch.getAddress().getStreet(), branch.getAddress().getHouse(),
        branch.getAddress().getOffice(), branch.getAddress().getPhone(),
        branch.getAddress().getRaw());

    DayModel dayModelMonday = new DayModel(branch.getSchedule().getMonday().getId(),
        branch.getSchedule().getMonday().getStart(), branch.getSchedule().getMonday().getEnd(),
        branch.getSchedule().getMonday().getBreakStart(),
        branch.getSchedule().getMonday().getBreakEnd(),
        branch.getSchedule().getMonday().isDayOff());
    DayModel dayModelTuesday = new DayModel(branch.getSchedule().getTuesday().getId(),
        branch.getSchedule().getTuesday().getStart(), branch.getSchedule().getTuesday().getEnd(),
        branch.getSchedule().getTuesday().getBreakStart(),
        branch.getSchedule().getTuesday().getBreakEnd(),
        branch.getSchedule().getTuesday().isDayOff());
    DayModel dayModelWednesday = new DayModel(branch.getSchedule().getWednesday().getId(),
        branch.getSchedule().getWednesday().getStart(),
        branch.getSchedule().getWednesday().getEnd(),
        branch.getSchedule().getWednesday().getBreakStart(),
        branch.getSchedule().getWednesday().getBreakEnd(),
        branch.getSchedule().getWednesday().isDayOff());
    DayModel dayModelThursday = new DayModel(branch.getSchedule().getThursday().getId(),
        branch.getSchedule().getThursday().getStart(), branch.getSchedule().getThursday().getEnd(),
        branch.getSchedule().getThursday().getBreakStart(),
        branch.getSchedule().getThursday().getBreakEnd(),
        branch.getSchedule().getThursday().isDayOff());
    DayModel dayModelFriday = new DayModel(branch.getSchedule().getFriday().getId(),
        branch.getSchedule().getFriday().getStart(), branch.getSchedule().getFriday().getEnd(),
        branch.getSchedule().getFriday().getBreakStart(),
        branch.getSchedule().getFriday().getBreakEnd(),
        branch.getSchedule().getFriday().isDayOff());
    DayModel dayModelSaturday = new DayModel(branch.getSchedule().getSaturday().getId(),
        branch.getSchedule().getSaturday().getStart(), branch.getSchedule().getSaturday().getEnd(),
        branch.getSchedule().getSaturday().getBreakStart(),
        branch.getSchedule().getSaturday().getBreakEnd(),
        branch.getSchedule().getSaturday().isDayOff());
    DayModel dayModelSunday = new DayModel(branch.getSchedule().getSunday().getId(),
        branch.getSchedule().getSunday().getStart(), branch.getSchedule().getSunday().getEnd(),
        branch.getSchedule().getSunday().getBreakStart(),
        branch.getSchedule().getSunday().getBreakEnd(),
        branch.getSchedule().getSunday().isDayOff());

    ScheduleModel scheduleModel =
        new ScheduleModel(branch.getSchedule().getId(), dayModelMonday, dayModelTuesday,
            dayModelThursday, dayModelWednesday, dayModelFriday, dayModelSaturday, dayModelSunday,
            branch.getSchedule().getRaw());

    return new BranchModel(branch.getId(), branch.getName(), addressModel, scheduleModel);
  }

  public List<BranchModel> transform(List<Branch> branchList) {
    List<BranchModel> branchModelList = new ArrayList<>(branchList.size());
    BranchModel branchModel;
    for (Branch branch : branchList) {
      branchModel = transform(branch);
      if (branchModel != null) {
        branchModelList.add(branchModel);
      }
    }

    return branchModelList;
  }
}

File: app/src/main/java/md/fusionworks/lifehack/data/api/model/exchange_rates/mapper/BranchDataMapper.kt
package md.fusionworks.lifehack.data.api.model.exchange_rates.mapper

import java.util.ArrayList
import md.fusionworks.lifehack.data.api.model.exchange_rates.Branch
import md.fusionworks.lifehack.ui.exchange_rates.model.AddressModel
import md.fusionworks.lifehack.ui.exchange_rates.model.BranchModel
import md.fusionworks.lifehack.ui.exchange_rates.model.DayModel
import md.fusionworks.lifehack.ui.exchange_rates.model.DistrictModel
import md.fusionworks.lifehack.ui.exchange_rates.model.LocalityModel
import md.fusionworks.lifehack.ui.exchange_rates.model.LocationModel
import md.fusionworks.lifehack.ui.exchange_rates.model.ScheduleModel

/**
 * Created by ungvas on 2/10/16.
 */
class BranchDataMapper {

  fun transform(branch: Branch): BranchModel {

    val localityModel = LocalityModel(branch.address!!.district!!.locality!!.id,
        branch.address!!.district!!.locality!!.name,
        branch.address!!.district!!.locality!!.code)
    val districtModel = DistrictModel(branch.address!!.district!!.id,
        branch.address!!.district!!.name, localityModel)
    val addressModel = AddressModel(
        LocationModel(branch.address!!.location!!.lat,
            branch.address!!.location!!.lng), branch.address!!.id, districtModel,
        branch.address!!.street, branch.address!!.house,
        branch.address!!.office, branch.address!!.phone,
        branch.address!!.raw)

    val dayModelMonday = DayModel(branch.schedule!!.monday!!.id,
        branch.schedule!!.monday!!.start, branch.schedule!!.monday!!.end,
        branch.schedule!!.monday!!.breakStart,
        branch.schedule!!.monday!!.breakEnd,
        branch.schedule!!.monday!!.isDayOff)
    val dayModelTuesday = DayModel(branch.schedule!!.tuesday!!.id,
        branch.schedule!!.tuesday!!.start, branch.schedule!!.tuesday!!.end,
        branch.schedule!!.tuesday!!.breakStart,
        branch.schedule!!.tuesday!!.breakEnd,
        branch.schedule!!.tuesday!!.isDayOff)
    val dayModelWednesday = DayModel(branch.schedule!!.wednesday!!.id,
        branch.schedule!!.wednesday!!.start,
        branch.schedule!!.wednesday!!.end,
        branch.schedule!!.wednesday!!.breakStart,
        branch.schedule!!.wednesday!!.breakEnd,
        branch.schedule!!.wednesday!!.isDayOff)
    val dayModelThursday = DayModel(branch.schedule!!.thursday!!.id,
        branch.schedule!!.thursday!!.start, branch.schedule!!.thursday!!.end,
        branch.schedule!!.thursday!!.breakStart,
        branch.schedule!!.thursday!!.breakEnd,
        branch.schedule!!.thursday!!.isDayOff)
    val dayModelFriday = DayModel(branch.schedule!!.friday!!.id,
        branch.schedule!!.friday!!.start, branch.schedule!!.friday!!.end,
        branch.schedule!!.friday!!.breakStart,
        branch.schedule!!.friday!!.breakEnd,
        branch.schedule!!.friday!!.isDayOff)
    val dayModelSaturday = DayModel(branch.schedule!!.saturday!!.id,
        branch.schedule!!.saturday!!.start, branch.schedule!!.saturday!!.end,
        branch.schedule!!.saturday!!.breakStart,
        branch.schedule!!.saturday!!.breakEnd,
        branch.schedule!!.saturday!!.isDayOff)
    val dayModelSunday = DayModel(branch.schedule!!.sunday!!.id,
        branch.schedule!!.sunday!!.start, branch.schedule!!.sunday!!.end,
        branch.schedule!!.sunday!!.breakStart,
        branch.schedule!!.sunday!!.breakEnd,
        branch.schedule!!.sunday!!.isDayOff)

    val scheduleModel = ScheduleModel(branch.schedule!!.id, dayModelMonday, dayModelTuesday,
        dayModelThursday, dayModelWednesday, dayModelFriday, dayModelSaturday, dayModelSunday,
        branch.schedule!!.raw)

    return BranchModel(branch.id, branch.name, addressModel, scheduleModel)
  }

  fun transform(branchList: List<Branch>) = branchList.map { transform(it) }
}
